### ISSUE TITLE
Feature: Warn about unused/unrecognized variables

### ISSUE THREAD

**Issue by mattbrictson** [capistrano/capistrano#1626]
The entire Capistrano configuration process hinges on declarations like these:

```ruby
set :some_special_name, "value"
```

This is simple, but the problem is there is no guidance (other than documentation or tutorials found via Google searches, which are often outdated and incorrect) in terms of which variables have special meaning.

It is easy to misspell a variable and then spend hours wondering why a deployment is failing. Unlike a Ruby method invocation or symbol lookup, there is no equivalent of `NoMethodError` or `NameError` when you make a mistake. When learning a new tool, this sort of feedback is incredibly helpful.

My proposal is for Capistrano to give users feedback when it thinks a variable has been mistyped, or when a variable no longer has any meaning (e.g `:copy_strategy` did something in Cap v2, but is meaningless in v3).

Of course this is easier said than done, and perhaps cannot be made to be 100% accurate, which is why it could be off by default, but turned on when:

* An uncaught exception is raised (e.g. deployment failed), in which case part of the error detail printed to the log could contain this variable information.
* Or a command-line switch could turn it on, like `--validate` or `--warnings`.

In terms of implementation, it could work like this:

1. Variables that are set via `set_default` are whitelisted (i.e. how Capistrano's own special variables and plugin's variables are initialized).
2. Other variables (i.e. those set in `deploy.rb` or in a stage file) generate warnings if they are `set` but subsequently never `fetch`ed.

Again, this could generate false negatives, but I think overall it would be useful to novice to intermediate users of Capistrano, where it is unlikely they are creating variables of their own, and therefore any `set` should be for a well-known variable name provided by the framework or plugins.

**Comment by leehambley** [comment]
Sounds worthwhile for sure, I'm just not sure how we can cover it completely. (Probably, we don't have to)

I wonder if we can also get away with a blacklist of known bad variables, as a quick intermediate step, and "tracking" of usage of other things separately.

**Comment by mattbrictson** [comment]
What about a "strict mode" where you aren't allowed to assign a value (`set`) unless it has been "declared" first via `set_default`? We could enable strict mode by default for newly capified projects (e.g. via `require "capistrano/strict"` in the generated Capfile), and it could be toggled for existing projects using a `--strict` command-line option. Too ambitious?

**Comment by leehambley** [comment]
I feel like unless any two plugins share a variable, it should never have
been allowed to be `set` at all, instead requiring some method call (which
could be rdoc'd, etc) on a plugin object.... I think that ship sailed in
2008 though, a strict mode doesn't sound too bad, and using `set_default`
for a plugin to name it's defaults doesn't sound horrible, but we'd have
*so* much ground to cover.

Lee Hambley
http://lee.hambley.name/
+49 (0) 170 298 5667

**Comment by mattbrictson** [comment]
You're right; existing plugins don't really use `set_default`, so it would be a huge shift. We'll need to pick a better battle. :smile:

**Comment by mattbrictson** [comment]
OK, implementation "plan B":

1. In `setup.rb`, after invoking `load:defaults`, make a note of all variables that are defined at this point. This will be the union of Capistrano's defaults plus any provided by plugins.
2. Proceed to load `deploy.rb` and the stage file.
3. Check again what variables are defined. Subtract the variable names found in step 1. The remainder are "suspicious".
4. ???

Haven't thought it all through, but maybe that's a start.

**Comment by leehambley** [comment]
Sounds like a decent start, at least I can't see any glaring holes. This
also reminds me that it'll be *very* helpful for the SCM split, that we can
avoid people acidentally naming svn_* vars when using the Git plugin

Lee Hambley
http://lee.hambley.name/
+49 (0) 170 298 5667

**Comment by leehambley** [comment]
I like that a lot. It works well for homebrew, and I'm a pretty seasoned
homebrew user, I still find use for that quite more often than I'd care to
admit.

Lee Hambley
http://lee.hambley.name/
+49 (0) 170 298 5667

**Comment by will-in-wi** [comment]
I like this a lot. It also gives a place to put other checks for frequent issues.

:+1:

**Comment by mattbrictson** [comment]
OK, here is actual output from my latest working proof-of-concept.

As you can see, the variables listing is a bit overwhelming, but perhaps that is unavoidable.

```
$ cap production doctor

Gems
====
capistrano         3.4.0
airbrussh          1.0.0
rake               10.5.0 (update available)
sshkit             1.9.0
capistrano-bundler 1.1.4
capistrano-rails   1.1.6

Variables
=========
:application                 "my-app"
:assets_prefix               "assets"
:assets_roles                [:web]
:branch                      "master"
:bundle_bins                 ["gem", "rake", "rails"]
:bundle_binstubs             false
:bundle_flags                "--deployment --retry=3"
:bundle_path                 #<Pathname:/home/deployer/apps/my-app/shared/bundle>
:bundle_roles                :all
:bundle_servers              [#<Capistrano::Configuration::Server:0x007f99e5c88490 @user="deployer", @hostname="REDACTED", @port=nil, @properties=#<Capistrano::Configuration::Server::Properties:0x007f99e5c88008 @properties={}, @roles=#<Set: {:app, :backup, :cron, :db, :redis, :sidekiq, :web}>>>]
:bundle_without              "development test"
:conditionally_migrate       false
:copy_strategy               :scp
:default_env                 {}
:deploy_to                   "/home/deployer/apps/my-app"
:format                      :airbrussh
:git_environmental_variables {:git_askpass=>"/bin/echo", :git_ssh=>"/tmp/my-app/git-ssh.sh"}
:keep_releases               10
:linked_dirs                 ["public/assets", ".bundle", "log", "tmp/pids", "tmp/cache", "tmp/sockets", "public/system", "public/assets"]
:linked_files                [".env.production", "config/database.yml", "config/unicorn.rb"]
:local_user                  "mbrictson"
:log_level                   :debug
:migration_role              :app
:pty                         false
:rails_env                   :production
:repo_url                    "git@bitbucket.org:REDACTED.git"
:scm                         :git
:ssh_options                 {:compression=>false, :keepalive=>true}
:stage                       :production
:tmp_dir                     "/tmp"

:copy_strategy is not a recognized Capistrano setting (config/deploy.rb:14)
```

Any suggestions for how to clean this up or make it better?

**Comment by leehambley** [comment]
> Any suggestions for how to clean this up or make it better?

Definitely print this as "markdown", maybe we can't use fenced codeblocks, but try and make sure if you copy it verbatim that pasting it in a GitHub issue, or on SO would render it as code where suitable. 

I think that means a fairly conservative 4 space indentation

**Comment by mattbrictson** [comment]
> make sure if you copy it verbatim that pasting it in a GitHub issue, or on SO would render it as code where suitable

Agreed, this is a great idea.

**Comment by mattbrictson** [comment]
Implemented in #1642.
